src/Utilities/FNADllMap.cs

```diff
@@ -0,0 +1,152 @@
+#region License
+/* FNA - XNA4 Reimplementation for Desktop Platforms
+ * Copyright 2009-2021 Ethan Lee and the MonoGame Team
+ *
+ * Released under the Microsoft Public License.
+ * See LICENSE for details.
+ */
+#endregion
+
+#if NET5_0
+
+#region Using Statements
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using System.Xml;
+#endregion
+
+namespace Microsoft.Xna.Framework
+{
+	internal static class FNADllMap
+	{
+		#region Private Static Variables
+
+		private static Dictionary<string, string> mapDictionary
+			= new Dictionary<string, string>();
+
+		#endregion
+
+		#region Private Static Methods
+
+		private static string GetPlatformName()
+		{
+			if (OperatingSystem.IsWindows())
+			{
+				return "windows";
+			}
+			else if (OperatingSystem.IsMacOS())
+			{
+				return  "osx";
+			}
+			else if (OperatingSystem.IsLinux())
+			{
+				return "linux";
+			}
+			else if (OperatingSystem.IsFreeBSD())
+			{
+				return "freebsd";
+			}
+			else
+			{
+				// Maybe this platform statically links?
+				return "unknown";
+			}
+		}
+
+		#endregion
+
+		#region DllImportResolver Callback Method
+
+		private static IntPtr MapAndLoad(
+			string libraryName,
+			Assembly assembly,
+			DllImportSearchPath? dllImportSearchPath
+		) {
+			string mappedName;
+			if (!mapDictionary.TryGetValue(libraryName, out mappedName))
+			{
+				mappedName = libraryName;
+			}
+			return NativeLibrary.Load(mappedName, assembly, dllImportSearchPath);
+		}
+
+		#endregion
+
+		#region Module Initializer
+
+		[ModuleInitializer]
+		public static void Init()
+		{
+			// Get the platform and architecture
+			string os = GetPlatformName();
+			string cpu = RuntimeInformation.ProcessArchitecture.ToString().ToLowerInvariant();
+
+			// Locate the config XML
+			Assembly assembly = Assembly.GetCallingAssembly();
```

Comment:
That would violate the original dllmap spec - this may be a case where behavior will diverge between JIT and AOT. Does AOT simply return null or does it throw an Exception?

Reply:
You can also use `AppContext.BaseDirectory`.

Reply:
Few attemps compiled with bflat:

## Assembly.GetCallingAssembly()
```csharp
using System;

var asm = System.Reflection.Assembly.GetCallingAssembly();
			
if (asm == null)
{
	Console.WriteLine("asm is null");
}
else
{
	Console.WriteLine(asm);
}
```

produces the following result:

```
Unhandled Exception: System.PlatformNotSupportedException: Operation is not supported on this platform.
   at System.Reflection.Assembly.GetCallingAssembly() + 0x5e
   at Test.Program.Main(String[]) + 0x9
   at Test!<BaseAddress>+0x2ae97a
```

---


## Environment.CurrentDirectory

```csharp
using System;


Console.WriteLine("Environment.CurrentDirectory = {0}", Environment.CurrentDirectory);
```


produces:

```
Environment.CurrentDirectory = C:\<my_folder_path>\Test
```


---

## AppContext.BaseDirectory

[thanks @lithiumtoast i do not aware about this]

```csharp
using System;


Console.WriteLine("AppContext.BaseDirectory = {0}", AppContext.BaseDirectory);
```


produces [notice the final backslash]:

```
AppContext.BaseDirectory = C:\<my_folder_path>\Test\ 
```

Reply:
Very well - we'll need to try GetCallingAssembly first, then catch and fall back to AppContext.BaseDirectory for safety.

Reply:
Hmm, so how are you supposed to use NativeLibrary with AOT? [SetDllImportResolver](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary.setdllimportresolver) takes an assembly parameter.

Reply:
It looks like Assembly.GetExecutingAssembly might work...? https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/src/tests/Interop/NativeLibrary/Callback/CallbackTests.cs#L36

Reply:
## Assembly.GetExecutingAssembly()
```csharp
using System;


Console.WriteLine("Assembly.GetExecutingAssembly = {0}", System.Reflection.Assembly.GetExecutingAssembly().Location);
```

### Minimum size
`bflat build --no-reflection --no-stacktrace-data --no-globalization --no-exception-messages --Os`

results:
```
Unhandled Exception: EETypeRva:0x00048E88: Arg_NotImplementedException
   at Test!<BaseAddress>+0xab100
   at Test!<BaseAddress>+0xab121
   at Test!<BaseAddress>+0x9da56
   at Test!<BaseAddress>+0xaa00a
   at Test!<BaseAddress>+0xaa3ac
   at Test!<BaseAddress>+0xb2c1b
   at Test!<BaseAddress>+0xb4520
   at Test!<BaseAddress>+0xbc589
```

### Standard build
`bflat build`

results [seems null/empty]:
```
Assembly.GetExecutingAssembly =
```

Reply:
Interesting, I just ran a non-bflat NativeAOT build to see what happens. GetExecutingAssembly returns the assembly information, but its Location is null/empty string. I'm kinda wondering if we should just hardcode "FNA.dll.config" for the filepath instead of using the Location string. I think it uses a relative file search path anyway... I can check that later

Reply:
A fallback to FNA.dll.config makes sense to me (maybe with a warning, even though it'd print 100% of the time on NativeAOT).

Reply:
[Assembly.Location will be always null/empty](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/il3000) when deployed as single file because the assembly doesn't exist on the filesystem anymore.

This is also true for CoreCLR single file publishing after .NET 5 (the single file that isn't just a glorified self-extracting ZIP) - it's not specific to NativeAOT single file.